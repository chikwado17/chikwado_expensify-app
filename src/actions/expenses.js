import database from '../firebase/firebase';


//creating action generator for addExpense
                                //creating default object for expenses then pass it to expenses object                        
export const addExpense = (expense) => ({
    type: 'ADD_EXPENSE',
    expense
});

//action dispatch of firebasre to redux 
export const startAddExpense = (expenseData = {}) => {
   
    return (dispatch) => {

        const {
            description = '',
            note = '',
            amount = 0,
            createdAt = 0
        } = expenseData;


        database.ref('expenses').push({ description,note,amount,createdAt }).then((ref)=>{
            //dispatching the function generator for database
            dispatch(addExpense({
                id:ref.key,   
                description,
                note,
                amount,
                createdAt
            }));
        });
    };
};



//creating action generator for removeExpense
export const removeExpense = ({id} = {}) => ({
    type: 'REMOVE_EXPENSE',
    id
});

//action generator for editExpense
export const editExpense = (id,updates) => ({
    type:'EDIT_EXPENSE',
    id,
    updates
});
//end of expenses reducer



//setup for firebase to fetch save data from database

//set_Expenses action generator
export const setExpenses = (expenses) => ({
    type:'SET_EXPENSES',
    expenses
});

//dispatching function for setExpenses for firebase database
export const startSetExpenses = () => {
    return (dispatch) => {
       return database.ref('expenses').once('value').then((snapshot) => {
            const expenses = [];

            snapshot.forEach((childSnapshot) => {
                    expenses.push({
                        id:childSnapshot.key,
                        ...childSnapshot.val()
                    });
            });

            dispatch(setExpenses(expenses));
        });
    }
}

