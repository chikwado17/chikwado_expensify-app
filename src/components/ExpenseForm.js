import React, { Component } from 'react';
import moment from 'moment';
import { SingleDatePicker } from 'react-dates';



class ExpenseForm extends Component {

        constructor(props){

            super(props);

            this.state = {
                //Checking if props.expense is exiting, if there is an expense already
                //edit it if there is none then show an empty string or empty input
                description: props.expense ? props.expense.description : '',
                note: props.expense ? props.expense.note : '',
                amount: props.expense ? (props.expense.amount / 100).toString() : '',
                createdAt: props.expense ? moment(props.expense.createdAt) :moment(),
                calenderFocused:false,
                error:''
            }
        }
        onDescriptionChange = (e) => {
            const description = e.target.value;

            this.setState(() => ({ description }));
        }

        onAddNote = (e) => {
            const note = e.target.value;

            this.setState(() => ({ note }))
        }
        onAmountChange = (e) => {
            const amount = e.target.value;

        if(!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)){

                this.setState(() => ({ amount }));
            }
        };
        onDateChange = (createdAt) => {
            if(createdAt){
                this.setState(() => ({ createdAt }));
            }
        };
        onFocusChange = ({focused}) => {
            this.setState(() => ({calenderFocused: focused}));
        }
        onSubmit = (e) => {
            //Stopping pagereload
            e.preventDefault();

            if(!this.state.description || !this.state.amount){
                this.setState(() => ({
                    error: 'Please Provide Description and Amount'
                }));
            }else{
                this.setState(() => ({
                    error:''
                }));
               
                //receiving onSubmit props down here from AddExpense
                //
                this.props.onSubmit({
                    description:this.state.description,
                    amount: parseFloat(this.state.amount, 10) * 100,
                    createdAt: this.state.createdAt.valueOf(),
                    note:this.state.note
                });
            }
        }
    render() { 
        return ( 

            <form className="form" onSubmit={this.onSubmit}>
                 {this.state.error && <p className="form__error">{this.state.error}</p>}
                    <input type="text"
                        placeholder="Description"
                        className="text-input"
                        autoFocus 
                            value={this.state.description}
                            onChange={this.onDescriptionChange}
                        />
                        
                        
                    <input type="number"
                     placeholder="Amount" 
                     className="text-input"
                     value={this.state.amount}
                     onChange={this.onAmountChange}
                     />

                    <SingleDatePicker
                            date={this.state.createdAt}
                            onDateChange={this.onDateChange}
                            focused={this.state.calenderFocused}
                            onFocusChange={this.onFocusChange}
                            numberOfMonths={1}
                            isOutsideRange={() => false}
                        />

                    <textarea
                    className="textarea"
                        cols="30"
                        rows="10"
                        placeholder="Add a note of your Expense"
                        value={this.state.note}
                        onChange={this.onAddNote}
                        >
                    </textarea>
                <div>
                    <input type="submit" value="Save Expense" className="button"/>
                </div>
            </form>
           
         );
    }
}
 
export default ExpenseForm;