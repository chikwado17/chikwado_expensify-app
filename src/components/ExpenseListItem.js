import React, { Component } from 'react';
import { Link } from 'react-router-dom';
import moment from 'moment';
import numeral from 'numeral';

                            //destructing ID and Dispatch as a props
const ExpenseListItem = ({ id, description,amount,createdAt}) => {
    return ( 
        <div>
            {/* Redirecting the links to edit page with the props ID*/}
           <Link to={`/edit/${id}`}>
                 <h3>{description}</h3>
           </Link>
           <p>
                {numeral(amount / 100).format('$0,0.00')} 
                -
                {moment(createdAt).format('MMMM Do, YYYY')}
            </p>
        </div>
     );
}

export default ExpenseListItem;