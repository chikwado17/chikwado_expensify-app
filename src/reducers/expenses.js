// creating a default value for expensesReducer state
const expensesReducerDefault = [];
//creating my expenseReducer
const expensesReducer = (state = expensesReducerDefault,action) => {
    switch(action.type){
        case 'ADD_EXPENSE':
        //using spread operator to add the current expense value to the old expense value without changing the expense object
            return[
                ...state,
                action.expense
            ];
        case 'REMOVE_EXPENSE':
            //removing the expenseOne object your the id, if the id is equal to the expense
            return state.filter(({id}) =>  id !== action.id);
            //updating the match with is the amount using map function to check through the object
        case 'EDIT_EXPENSE':
            return state.map((expense) => {
                if(expense.id === action.id){
                    return {
                    ...expense,
                        //finding the update from action.update which is coming from the exitExpense action generator
                    ...action.updates
                    };
                }else{
                    return expense;
                };
            }); 
        case 'SET_EXPENSES':
            return action.expenses
        default:
            return state;
    }
}
//End of expenseReducer

export default expensesReducer;