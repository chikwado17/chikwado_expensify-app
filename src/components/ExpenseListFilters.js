import React from 'react';
import { connect } from 'react-redux';
import { setTextFilter } from '../actions/filters';
import { sortByAmount } from '../actions/filters';
import { sortByDate, setStartDate, setEndDate } from '../actions/filters';
import { DateRangePicker } from 'react-dates';




class ExpenseListFilters extends React.Component {

    state = {

        calenderFocused: null
    }


    onDatesChange = ({ startDate , endDate}) => {

        this.props.dispatch(setStartDate(startDate));
        this.props.dispatch(setEndDate(endDate));

    };

    onFocusChange = ( calenderFocused ) => {
        this.setState(() => ({ calenderFocused: calenderFocused }))
    }
    render() {

        return ( 
            <div className="content-container">
                <div className="input-group">
                    <div className="input-group__item">
                         {/*/setting the filters text to the input value using props.filters.text*/}
                        <input type="text" placeholder="search expenses" className="text-input" value={this.props.filters.text} onChange={(e) => {
                            //dispaction an action generator
                            this.props.dispatch(setTextFilter(e.target.value));
                        }}/>
                    </div>
                    <div className="input-group__item">
                         {/* sorting by date and amount using sortBy then dispatch sortbyDate and sortByAmount */}
                            <select value={this.props.filters.sortBy} className="select" onChange={(e) => {
                                
                                if(e.target.value === 'date'){
                                this.props.dispatch(sortByDate());
                                }else if(e.target.value === 'amount'){
                                    this.props.dispatch(sortByAmount());
                                }
                                
                                }}>
                
                                <option value="date">Date</option>
                                <option value="amount">Amount</option>
                            </select>
                    </div>
                    <div className="input-group__item">
                        <DateRangePicker
                            startDate={this.props.filters.startDate}
                            endDate={this.props.filters.endDate}
                            onDatesChange={this.onDatesChange}
                            focusedInput={this.state.calenderFocused}
                            onFocusChange={this.onFocusChange}
                            showClearDates={true}
                            isOutsideRange={ () => false}
                            numberOfMonths={1}
                        />
                    </div>
                </div>
            </div>
         );
    }
}




//HOC
const mapStateToProps = (state) => {
    return {
        filters:state.filters
    }
}
export default connect(mapStateToProps)(ExpenseListFilters);