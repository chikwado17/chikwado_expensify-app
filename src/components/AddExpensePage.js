import React, { Component } from 'react';
import ExpenseForm from './ExpenseForm';
import  { connect }  from 'react-redux';
import { startAddExpense } from '../actions/expenses';


const AddExpensePage = (props) => {
    return ( 
        <div>
            <div className="page-header">
                <div className="content-container">
                        <h1 className="page-header__title">Add Expense</h1>
                </div>
            </div>

             {/* Passing down onSubmit as a props down to ExpenseForm */}
            <div className="content-container">
                <ExpenseForm 
                        onSubmit={(expense) => {
                            props.dispatch(startAddExpense(expense))
                            //redirecting Url to HomePage
                            props.history.push('/')
                        }}
                    />
            </div>
        </div>
     );
}

export default connect()(AddExpensePage);