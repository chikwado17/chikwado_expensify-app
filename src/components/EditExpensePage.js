import React, { Component } from 'react';
import { connect } from 'react-redux';
import ExpenseForm from './ExpenseForm';
import { startEditExpense, startRemoveExpense } from '../actions/expenses';



const EditExpensePage = (props) => {
    return ( 
        <div>
            <ExpenseForm 
               // {/* passing the expenses down as a props to ExpenseForm component and make an update to the state*/}
                expense={props.expense}

                onSubmit={(expense) => {
                    //the editExpense receives two argument which is props.expense.id -> (the ID of the particular expense to edit) and expense
                   props.dispatch(startEditExpense(props.expense.id,expense));
                   props.history.push('/')
                }}
            />

             <button onClick={() =>
               {
                   //dispatching here is a props and we are removing an expense here by an ID passing it expense ID using props
                   props.dispatch(startRemoveExpense({ id: props.expense.id}));
                   props.history.push('/')
               }}
               >Remove</button>
               
        </div>
     );
}



const mapStateToProps = (state, props) =>{
    return {
        expense:state.expenses.find((expense) => {
            //if expenses ID is equal to edit URL ID then pick the expenses
            return expense.id === props.match.params.id
        })
    }
}

export default connect(mapStateToProps)(EditExpensePage);