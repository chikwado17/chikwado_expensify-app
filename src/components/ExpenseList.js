 import React, { Component } from 'react';
 import { connect } from 'react-redux';
 import ExpenseListItem from './ExpenseListItem';
 import getVisibleExpenses from '../selectors/expenses'; 

 export const ExpenseList = (props) => {
     return ( 
         <div>
            {
                props.expenses.length === 0 ? 
                (<p>No Expenses</p>) : 
                (
                    props.expenses.map((expense) => {
                        return <ExpenseListItem key={expense.id} {...expense} />
                    })
                )
            }
         </div>
      )
 }



//HOC 
 const mapStateToProps = (state) => {
     return {
        expenses:getVisibleExpenses(state.expenses,state.filters)
     }
 }
 export default connect(mapStateToProps)(ExpenseList);
 //End of HOC