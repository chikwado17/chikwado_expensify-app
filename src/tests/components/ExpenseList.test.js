import React from 'react';
import { shallow } from 'enzyme';
import { ExpenseList } from '../../components/ExpenseList';
import expenses from '../fixtures/expenses';



test("should test ExpenseList with expense", () => {

    const wrapper = shallow(<ExpenseList expenses={expenses}/>);
    expect(wrapper).toMatchSnapshot();
});

test("should test ExpenseList with an empty array", () => {

    const wrapper = shallow(<ExpenseList expenses={[]}/>);
    expect(wrapper).toMatchSnapshot();
});