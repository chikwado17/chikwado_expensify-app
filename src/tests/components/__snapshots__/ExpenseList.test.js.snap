// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should test ExpenseList with an empty array 1`] = `
<div>
  <p>
    No Expenses
  </p>
</div>
`;

exports[`should test ExpenseList with expense 1`] = `
<div>
  <ExpenseListItem
    amount={195}
    createdAt={0}
    description="uber"
    id="1"
    key="1"
    note=""
  />
  <ExpenseListItem
    amount={109500}
    createdAt={-345600000}
    description="Rent"
    id="2"
    key="2"
    note=""
  />
  <ExpenseListItem
    amount={100}
    createdAt={345600000}
    description="ubr"
    id="3"
    key="3"
    note=""
  />
</div>
`;
